modules/b2b_logic/logic.c

```diff
@@ -473,9 +474,25 @@ int process_bridge_dialog_end(b2bl_tuple_t* tuple, unsigned int hash_index,
 		}
 		else
 		{
-			/* the entity to connect replied with negative reply */
-			b2b_end_dialog(tuple->bridge_entities[0], tuple, hash_index);
-			b2b_mark_todel(tuple);
+			if(tuple->bridge_flags & B2BL_BR_FLAG_RETURN_AFTER_FAILURE)
+			{
+				/* Delete failed entity */
+				b2bl_delete_entity(bentity, tuple, hash_index, 1);
+
+				/* Restore initial bridge */
+				tuple->bridge_entities[1] = tuple->bridge_entities[0];
+				tuple->bridge_entities[0] = tuple->servers[0];
```

Comment:
Same idea as with the refer sending entity, we cannot assume that this is always a server entity.